dask/base.py

```diff
@@ -684,6 +684,34 @@ def _normalize_function(func):
             return str(func)
 
 
+def _hash_numpy_data(x):
+    try:
+        # string fast-path
+        return hash_buffer_hex('-'.join(x.flat).encode(encoding='utf-8', errors='surrogatepass'))
+    except UnicodeDecodeError:
+        # bytes fast-path
+        return hash_buffer_hex(b'-'.join(x.flat))
+    except TypeError:
+        # object slow-path
+        def to_bytes(item):
+            try:
+                # string item
+                return item.encode(encoding='utf-8', errors='surrogatepass')
+            except UnicodeDecodeError:
+                # byte item
+                return item
+            except AttributeError:
+                # object item
+                return unicode(item).encode(encoding='utf-8', errors='surrogatepass')
+
+        return hash_buffer_hex(
+            b'-'.join(
+                to_bytes(item)
+                for item in x.flat
+            )
+        )
```

Comment:
I wonder if it would be safe to rely on pickle here instead.

```python
hash_buffer_hex(pickle.dumps(x))
```

This might be simpler and faster . I *think* it should also be safe.

If we go with this route (happy to hear pushback) then I'd prefer to keep the code inline just to avoid the indirection.

Reply:
So it depends on our assumption:

1. **Is every object pickable?**
    This is by far not the default. If not, we would need to implement a cloudpickle fallback I think, similar to what we do for remote/distributed execution. Or do we want pickle + yet-another fallback?
2. **How deterministic is pickle?**
    If just a single bit flips, you get a different hash. So I think this boils down to the question how well-designed we think custom reductions/getstate implementations are. Also, how bad would it be if for some types we generate a different hash every time? Is this just an optimization or is it also correctness that we would violate?